lib/grammers-client/src/types/input_media.rs

```diff
@@ -0,0 +1,283 @@
+        self.media = Some(
+            (tl::types::InputMediaDocumentExternal {
+                spoiler: false,
+                url: url.into(),
+                ttl_seconds: self.media_ttl,
+            })
+            .into(),
+        );
+        self
+    }
+
+    /// Add additional attributes to the media.
+    ///
+    /// This must be called *after* setting a file.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # async fn f(client: &mut grammers_client::Client) -> Result<(), Box<dyn std::error::Error>> {
+    /// # let audio = client.upload_file("audio.flac").await?;
+    /// #
+    /// use std::time::Duration;
+    /// use grammers_client::{types::Attribute, InputMedia};
+    ///
+    /// let media = InputMedia::caption("").document(audio).attribute(
+    ///    Attribute::Audio {
+    ///        duration: Duration::new(123, 0),
+    ///        title: Some("Hello".to_string()),
+    ///        performer: Some("World".to_string()),
+    ///    }
+    /// );
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn attribute(mut self, attr: Attribute) -> Self {
+        if let Some(tl::enums::InputMedia::UploadedDocument(document)) = &mut self.media {
+            document.attributes.push(attr.into());
+        }
+        self
+    }
+
+    /// Copy media from an existing message.
+    ///
+    /// You can use this to send media from another message without re-uploading it.
+    pub fn copy_media(mut self, media: &Media) -> Self {
+        self.media = media.to_raw_input_media();
+        self
+    }
+
+    /// Include the uploaded file as a document file in the album.
+    ///
+    /// You can use this to send any type of media as a simple document file.
+    ///
+    /// The text will be the caption of the file, which may be empty for no caption.
+    pub fn file(mut self, file: Uploaded) -> Self {
+        let mime_type = self.get_file_mime(&file);
+        let file_name = file.name().to_string();
+        self.media = Some(
+            (tl::types::InputMediaUploadedDocument {
+                nosound_video: false,
+                force_file: true,
+                spoiler: false,
+                file: file.raw,
+                thumb: None,
+                mime_type,
+                attributes: vec![(tl::types::DocumentAttributeFilename { file_name }).into()],
+                stickers: None,
+                ttl_seconds: self.media_ttl,
+            })
+            .into(),
+        );
+        self
+    }
+
+    /// Change the media's Time To Live (TTL).
+    ///
+    /// For example, this enables you to send a `photo` that can only be viewed for a certain
+    /// amount of seconds before it expires.
+    ///
+    /// Not all media supports this feature.
+    ///
+    /// This method should be called before setting any media, else it won't have any effect.
+    pub fn media_ttl(mut self, seconds: i32) -> Self {
+        self.media_ttl = if seconds < 0 { None } else { Some(seconds) };
+        self
+    }
```

Comment:
Can this actually be used in albums? What happens if different photos have different TTLs?

Reply:
I think the server only uses the first media parameter.